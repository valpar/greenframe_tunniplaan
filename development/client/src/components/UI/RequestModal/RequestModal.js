import React, { Fragment, useEffect } from "react";
import ReactDOM from "react-dom";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faThumbsUp } from "@fortawesome/free-regular-svg-icons";
import { faExclamation } from "@fortawesome/free-solid-svg-icons";
import { Spinner } from "../Spinner";

const Backdrop = (props) => {
  useEffect(() => {
    const themeColorMeta = document.querySelector('meta[name="theme-color"]');

    themeColorMeta.setAttribute("content", "#5C5C5C");

    return () => {
      themeColorMeta.setAttribute("content", "#fff");
    };
  }, []);

  return (
    <div
      data-testid="backdrop"
      className="fixed top-0 left-0 w-full h-screen z-20 bg-neutral-900 opacity-70"
      onClick={props.onClose}
    />
  );
};

const RequestModalOverlay = (props) => {
  return (
    <div
      data-testid="modal-overlay"
      className={`fixed top-1/3 lg:top-1/4 w-11/12 lg:w-96 h-auto left-1/2 -translate-x-1/2 bg-white rounded z-40 shadow-lg animate-slideDown ${props.customStyle}`}
    >
      {props.loading && <Spinner containerStyle="p-12" />}
      {!props.loading && (
        <div className="flex justify-center px-8 pt-8 pb-4 text-xl">
          <p>{props.modalMessage}</p>
        </div>
      )}
      {props.success && (
        <>
          <div className=" flex justify-center text-collegeGreen text-5xl py-2">
            <FontAwesomeIcon icon={faThumbsUp} />
          </div>
          <div className="w-full h-8">
            <div className="h-1 w-0 bg-collegeGreen animate-greenLoader"></div>
          </div>
        </>
      )}

      {props.error && (
        <>
          <div className="flex justify-center pb-4 text-4xl lg:text-6xl text-collegeRed">
            <FontAwesomeIcon icon={faExclamation} />
          </div>
          <div className="flex justify-between px-4 pb-4">
            <button
              onClick={props.onDecline}
              type="button"
              className="px-4 py-2 w-28 border border-borderGray text-sm shadow hover:bg-borderGray hover:shadow-lg duration-150"
            >
              KATKESTA
            </button>
            <button
              onClick={props.onConfirm}
              type="button"
              className="px-4 py-2 w-28 border border-borderGray text-sm shadow hover:bg-borderGray hover:shadow-lg duration-150"
            >
              UUESTI
            </button>
          </div>
        </>
      )}
    </div>
  );
};

const portalElement = document.getElementById("overlays");

const RequestModal = (props) => {
  return (
    <Fragment>
      {ReactDOM.createPortal(<Backdrop />, portalElement)}
      {ReactDOM.createPortal(<RequestModalOverlay {...props} />, portalElement)}
    </Fragment>
  );
};

export default RequestModal;
