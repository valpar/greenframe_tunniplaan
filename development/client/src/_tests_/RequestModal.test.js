import React from 'react';
import { render, fireEvent, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import RequestModal from '../components/UI/RequestModal/RequestModal';

beforeAll(() => {
    const meta = document.createElement('meta');
    meta.setAttribute('name', 'theme-color');
    document.head.appendChild(meta);
  });
  
afterAll(() => {
  // eslint-disable-next-line testing-library/no-node-access
  document.head.removeChild(document.querySelector('meta[name="theme-color"]'));
});

// kui seda ei tee (createPortal), siis target "is not DOM element"
jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  createPortal: (element, target) => element, 
}));

describe('Modal', () => {
  it("matches snapshot", () => {
    const tree = renderer.create(<RequestModal />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('renders with Backdrop and ModalOverlay', () => {
    render(<RequestModal />);
    expect(screen.getByTestId('modal-overlay')).toBeInTheDocument();
  });
  it('displays spinner when loading is true', () => {
    render(<RequestModal loading={true} />);
    expect(screen.getByTestId('spinner')).toBeInTheDocument();
  });
  it('calls onConfirm when the confirm button is clicked', () => {
    const handleConfirm = jest.fn();
    render(<RequestModal onConfirm={handleConfirm} error={true} />);
    
    fireEvent.click(screen.getByText('UUESTI'));
    expect(handleConfirm).toHaveBeenCalled();
  });
});