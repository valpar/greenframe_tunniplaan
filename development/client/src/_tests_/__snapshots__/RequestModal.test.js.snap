// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal matches snapshot 1`] = `
Array [
  <div
    className="fixed top-0 left-0 w-full h-screen z-20 bg-neutral-900 opacity-70"
    data-testid="backdrop"
  />,
  <div
    className="fixed top-1/3 lg:top-1/4 w-11/12 lg:w-96 h-auto left-1/2 -translate-x-1/2 bg-white rounded z-40 shadow-lg animate-slideDown undefined"
    data-testid="modal-overlay"
  >
    <div
      className="flex justify-center px-8 pt-8 pb-4 text-xl"
    >
      <p />
    </div>
  </div>,
]
`;
